### PR TITLE
MGMT-2456: new cluster installing-pending-user-action state

### PR DESCRIPTION
**Pull request by RazRegev**

### CODE REVIEW

**Review by avishayt — COMMENTED** [review]
Looks fine to me, need another review by Michael or Ori

**Comment by RazRegev** [comment]
/hold
waiting for ui changes to be merged first in order not to break the current flow.

File: internal/cluster/statemachine.go
Comment (eranco74) [reply internal/cluster/statemachine.go:188]: Good catch

File: internal/cluster/statemachine.go
Comment (eranco74) [inline internal/cluster/statemachine.go:38]: Curios, how this can happen?

**Review by eranco74 — COMMENTED** [review]
Looks good

File: internal/cluster/statemachine.go
Comment (avishayt) [reply internal/cluster/statemachine.go:38]: How can we get from Installing to Insufficient as well?

Comment (filanov) [reply internal/cluster/statemachine.go:38]: +1

Comment (RazRegev) [reply internal/cluster/statemachine.go:38]: during installation if some hosts are in error the ui exposes reset option directly.

Comment (RazRegev) [reply internal/cluster/statemachine.go:38]: @avishayt we can get to it by skipping the cancellation and resetting the cluster from installing

Comment (filanov) [reply internal/cluster/statemachine.go:38]: like a hacky tool to try to fix installation?

File: docs/ClusterStatus.dot
Comment (filanov) [inline docs/ClusterStatus.dot:35]: Please add the error case as well `"installing-pending-user-action" -> error`

File: internal/cluster/statemachine.go
Comment (filanov) [inline internal/cluster/statemachine.go:193]: This condition is not correct. we will fail the installation only if one of the masters is in error 
Take a look at `IsInstalling` condition.

File: internal/cluster/statemachine.go
Comment (filanov) [inline internal/cluster/statemachine.go:193]: i don't think that you need `stateswitch.And(` here because you have only one condition, `stateswitch.Not(th.IsInstalling)` is enough.

File: internal/cluster/statemachine.go
Comment (filanov) [inline internal/cluster/statemachine.go:227]: Just to be on the safe side i think that it should be 
```
		Condition: stateswitch.And(
			th.IsInstallingPendingUserAction,
			th.IsInstalling),
```
You can have host in error and `IsInstallingPendingUserAction`.

File: internal/cluster/statemachine.go
Comment (filanov) [inline internal/cluster/statemachine.go:266]: Not sure in which case `stateswitch.Not(th.IsInstallingPendingUserAction)` is needed, i think that for this transition is `stateswitch.And(stateswitch.Not(th.IsFinalizing), stateswitch.Not(th.IsInstalling))` covering all the cases 
if you are not finilazing and not installing then  it means that one of the masters in error, it's not relevant if any of the hosts is pending for user action or not. 
Unless i missed something

Comment (RazRegev) [reply internal/cluster/statemachine.go:266]: you are right, I missed that IsInstalling includes HostStatusInstallingPendingUserAction

File: internal/cluster/statemachine.go
Comment (RazRegev) [reply internal/cluster/statemachine.go:38]: yes, and also for cases that hosts in error, ui will expose reset button directly.

File: internal/host/transition.go
Comment (filanov) [inline internal/host/transition.go:455]: Not relevant for this stage.
It is used in a transition that validate start of the installation, that we started to run the installer on the host, there is no wrong boot order in this case.

File: internal/host/transition.go
Comment (filanov) [inline internal/host/transition.go:472]: IMO this validation should be relevant for specific stages of the installation, for example writing to disk can still timeout and it's not related to wrong boot order or any user action. So maybe need to check the stage as well. 
waiting for control plane can be affected by this state.
@avishayt @eranco74 what do you think?

Comment (avishayt) [reply internal/host/transition.go:472]: I don't understand.  How can the host be in pending-user-action while writing to disk?  It should only happen in the reboot stage.

Comment (filanov) [reply internal/host/transition.go:472]: This is exactly what i'm saying, this validation should be done on specific stages of the installation.

Comment (avishayt) [reply internal/host/transition.go:472]: As an optimization?

Comment (filanov) [reply internal/host/transition.go:472]: No, it will hide bugs, or at least delay detection.

Comment (avishayt) [reply internal/host/transition.go:472]: I don't see how but I may be missing some nuance here.  I assumed that if we're in a different stage, then we're not in pending-user-action state, and the check will return false.

Comment (filanov) [reply internal/host/transition.go:472]: waiting for user action can happen if one of the hosts is having a wrong boot order. So if it's a master then other hosts will not be able to connect to control plane.
This validation checking only cluster state, and not the current stage of the host. 
So the validation should be that cluster is in `IsInstallationPendingUserAction` and the current installation stage is one of 
1. Waiting for control plane
2. Start waiting for control plane
3. Rebooting

File: internal/host/transition.go
Comment (filanov) [inline internal/host/transition.go:466]: Shouldn't it be true?

**Review by filanov — APPROVED** [review]

**Review by filanov — APPROVED** [review]

**Pull request merged**
